pkg/tagset/slice_builder.go

```diff
@@ -0,0 +1,186 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.Datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package tagset
+
+import (
+	"github.com/twmb/murmur3"
+)
+
+// SliceBuilder is used to build tagsets tag-by-tag, before "freezing" into a Tags instance. SliceBuilders are not threadsafe.
+//
+// A SliceBuilder goes through three stages in its lifecycle:
+//
+//     1. adding tags (begins on call to factory.NewSliceBuilder).
+//     2. frozen (begins on call to bldr.FreezeSlice).
+//     3. closed (begins on call to bldr.Close).
+//
+// The Add* methods may only be called in the "adding tags" stage. No methods
+// may be called in the "closed" stage.
+//
+// In general, the easiest way to ensure these stages are followed (and allow reviewers
+// to verify this) is to use a builder in a single method, as shown in the example.  Avoid
+// storing builders in structs.
```

Comment:
What do you think about adding an explanation that this struct stores tags by level and allows to merge them using a level range?

Reply:
Ah, good point, thanks!